nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/services/azure/keyvault/AzureKeyVaultClientService.java

```diff
@@ -0,0 +1,242 @@
+                            return getSecretFromKeyVault(secretName);
+                        }
+                    });
+        } else {
+            secretCache = null;
+            logger.info("Secret cache disabled because cache size is set to 0");
+        }
+    }
+
+    static List<PropertyDescriptor> descriptors = new ArrayList<>();
+
+    static {
+        descriptors.add(AzureKeyVaultUtils.KEYVAULT_NAME);
+        descriptors.add(AzureKeyVaultUtils.SP_CLIENT_ID);
+        descriptors.add(AzureKeyVaultUtils.SP_CLIENT_SECRET);
+        descriptors.add(AzureKeyVaultUtils.TENANT_ID);
+        descriptors.add(AzureKeyVaultUtils.ENDPOINT_SUFFIX);
+        descriptors.add(AzureKeyVaultUtils.USE_MANAGED_IDENTITY);
+        descriptors.add(AzureKeyVaultUtils.CACHE_SIZE);
+        descriptors.add(AzureKeyVaultUtils.CACHE_TTL_AFTER_WRITE);
+    }
+
+    @Override
+    protected List<PropertyDescriptor> getSupportedPropertyDescriptors() {
+        return descriptors;
+    }
+
+    @Override
+    protected Collection<ValidationResult> customValidate(ValidationContext validationContext) {
+        final List<ValidationResult> results = new ArrayList<>();
+
+        final String keyVaultName = validationContext.getProperty(
+                AzureKeyVaultUtils.KEYVAULT_NAME).getValue();
+        final String clientID = validationContext.getProperty(
+                AzureKeyVaultUtils.SP_CLIENT_ID).getValue();
+        final String clientSecret = validationContext.getProperty(
+                AzureKeyVaultUtils.SP_CLIENT_SECRET).getValue();
+        final String tenantID = validationContext.getProperty(
+                AzureKeyVaultUtils.TENANT_ID).getValue();
+        final Boolean useManagedIdentity = validationContext.getProperty(
+                AzureKeyVaultUtils.USE_MANAGED_IDENTITY).asBoolean();
+
+        if (StringUtils.isBlank(keyVaultName)) {
+            results.add(new ValidationResult.Builder()
+                    .subject(this.getClass().getSimpleName())
+                    .valid(false)
+                    .explanation(AzureKeyVaultUtils.KEYVAULT_NAME.getDisplayName() +" is required")
+                    .build());
+        } else if (useManagedIdentity && (StringUtils.isNotBlank(clientID)
+                || StringUtils.isNotBlank(clientSecret))) {
+            results.add(new ValidationResult.Builder()
+                    .subject(this.getClass().getSimpleName())
+                    .valid(false)
+                    .explanation("if " + AzureKeyVaultUtils.USE_MANAGED_IDENTITY.getDisplayName()
+                            + " is used then " + AzureKeyVaultUtils.SP_CLIENT_ID.getDisplayName()
+                            + ", " + AzureKeyVaultUtils.SP_CLIENT_SECRET.getDisplayName()
+                            + ", " + AzureKeyVaultUtils.TENANT_ID.getDisplayName()
+                            + " should be blank.")
+                    .build());
+        } else if (!useManagedIdentity && (StringUtils.isBlank(clientID)
```

Comment:
PR #4843 introduced support for Service Principals and NIFI-8277 proposes support for Client Certificate credentials.  This method and `createKeyValueSecretClient()` would benefit from some refactoring along with the existing credential handling to avoid duplication and ensure that that new credential implementations would be supported.

Reply:
I am not sure regarding this, are you asking to refactor so that these service principal details could also be fetched from a keyvault

Reply:
This method, and the other existing approach to determining which type of client credential to use share some of the same logic, so it would be helpful to reduce the duplication down the road.  It sounds like it is better to wait until NIFI-8277 to look at refactoring.